scripts/cube_fx.lua

```diff
@@ -290,11 +300,59 @@ local function cube_victory(size, results)
   end
 end
 
+-- Map to determine if some machine is using the cube
+-- in an "efficient" manner. 
+local cube_machine_condition_handlers = {
+  ["status"] = function(entity, cube_type)
+    return entity.status == defines.entity_status.working
+  end,
+  ["handcrafting"] = function(character, cube_type)
+    -- We could search through the crafting queue, but this function
+    -- needs to be quick because it's called every 6 ticks. Lets
+    -- rather assume that if the player doesn't have it in their
+    -- inventory or cursor then the player is crafting with it.
+    -- That doesn't handle it being in queues though, but it's
+    -- better than nothing.
+    local player = character.player
+    if not player then return end
+    local cursor_stack = player.cursor_stack
+    if cursor_stack.valid and cursor_stack.valid_for_read then
+      if cursor_stack.name == cube_type then return false end
+    end
+    local inventory = player.get_main_inventory()
+    if inventory.find_item_stack(cube_type) then return false end
+    return true
+  end,
+}
+
+-- Map to determine in which machine types the cube can be
+-- seen as being "utilization". The value will show what 
+-- requirement this machine has to be "productive" with
+-- the cube. 
+local cube_working_machine_types = {
+  ["character"]           = cube_machine_condition_handlers["handcrafting"],
+  ["assembling-machine"]  = cube_machine_condition_handlers["status"],
+  ["furnace"]             = cube_machine_condition_handlers["status"],
+  ["boiler"]              = cube_machine_condition_handlers["status"],
+  ["generator"]           = cube_machine_condition_handlers["status"],
+  ["burner-generator"]    = cube_machine_condition_handlers["status"],
+  ["rocket-silo"]         = cube_machine_condition_handlers["status"],
+  ["reactor"]             = cube_machine_condition_handlers["status"], -- Does this work?
```

Comment:
I don't think this is required. We know:
 - If the entity is in a `assembly-machine`, `furnace` or `generator` is _has_ to be either an ingredient or the fuel.
 - If this is the case the machine can only be actively running if the cube is is being used. And in this case the status will always be `working`. All other cases have different statusses:
   - If it's an ingredient and it's waiting for other ingredients it will be `item_ingredient_shortage` or `no_ingredients`.
   - If it's in the output of a crafter it will be `full_output`.
   - If it's the fuel (furnace or burner), but it cannot be burned yet do to ingredient shortage, or the output being full, then the status will be what's mentioned above.
   - When in the output slot of the nuclear reactor it says `no fuel`
   
> This fixes some rare cases (e.g. cube sitting in the output slot of a fuel refinery crafting a non-cube recipe).

This is possible. If you change it to a non-cube recipe then the cube will be taken out of the entity. If the cube is in a crafter, then the crafter has to have a cube recipe set.
   
Only thing I can think of is maybe while launching the rocket, but I'm no t sure if the silo is used.

I might be missing some other edge case? But to me it seems very accurate and simple

Reply:
> If the entity is in a assembly-machine, furnace or generator is has to be either an ingredient or the fuel.

This isn't true. The player can place the cube wherever they want. They can put the cube in an output slot of a fabricator that happens to be crafting something else. There's no reason to do that and it's weird, but your code would count that as the cube being used.

Reply:
I'll have a look tonight if I can get it into a state like this :)

Reply:
> They can put the cube in an output slot of a fabricator that happens to be crafting something unrelated

I still don't think this is possible :) 

![image](https://github.com/grandseiken/factorio-ultracube/assets/39875036/4788183e-eaa2-4edb-8f57-adc13916c4c8)
![image](https://github.com/grandseiken/factorio-ultracube/assets/39875036/18245208-ac37-45eb-aba6-b58114713e06)

---

What's maybe possible is the following case where you put the cube in a furnace output slot. But when it's an `ultra-dense` then it's impossible that's it's currently burning, meaning status is always `No Fuel` as seen in this screen grab. One case where this isn't true is with the phantom cubes, by placing a phantom cube in the output slot, but that would mean that another phantom cube is burning, and this it's status would be `working` correctly.
![image](https://github.com/grandseiken/factorio-ultracube/assets/39875036/f270e917-2321-4f49-a3f0-0e835df39e2b)

---

I'm sorry if I'm coming across stubborn. I just feel that any other approach with specific tests for every entity type would introduce more potential bugs and edge cases, as well as being worse for performance. Whereas this approach is easy to understand, fast, and fool-proof.

That said, this is your code, and I will change it if you feel strongly about it :)

Reply:
Wait, one valid edge case is placing the cube in the boiler output slot, and then feeding the boiler with `Condensed Fuel`. I'll add a special case for the boiler.
![image](https://github.com/grandseiken/factorio-ultracube/assets/39875036/43b7d88f-e565-4154-b5d9-b44eea2d6fb5)

Reply:
Okay, added another category `crafter` that works for `assembling-machines`, `furnaces`, (and `silos`?). It checks if the cube is being burned, and then if it's an ingredient :)